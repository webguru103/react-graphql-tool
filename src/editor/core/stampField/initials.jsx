import * as React from 'react';
import classNames from 'classnames';
import { withStyles } from '@material-ui/core';
import CircularProgress from '@material-ui/core/CircularProgress';
import InitialIcon from '../../../assets/initial-alt.svg';
import { EDITOR_MODE } from '../../../constants';
import { STAMP_URL_PLACEHOLDER, DEFAULT_STAMP_URL_FONT_SIZE } from '../../constants';
import { getWidthHeightFromSVGString } from './stamp.utils';

type Props = {
  classes: Object,
  value: string,
  onClick: () => void,
  stampLookupUrl?: string,
  wipe: () => void,
  mode: $Keys<typeof EDITOR_MODE>,
  loading: boolean,
  height: number,
  width: number,
  zoom: number,
  color: { red: number, green: number, blue: number },
  disabled: boolean,
}

type State = {
  hovered: boolean,
}

const styles = theme => ({
  root: {
    minHeight: '100%',
    minWidth: '100%',
    display: 'flex',
    alignItems: 'center',
    justifyContent: 'center',
    flexFlow: 'column',
  },
  loaderContainer: {
    borderRadius: 5,
    backgroundColor: theme.palette.primary.main,
  },
  loader: {
    color: 'white',
  },
  signButton: {
    border: 'none',
    borderRadius: 5,
    color: 'white',
    backgroundColor: 'transparent',
    outline: 'none',
  },
  signText: {
    fontSize: 12,
  },
  initialsImage: {
    outline: 'none',
    maxWidth: '100%',
    maxHeight: '100%',
    width: 'inherit',
    height: 'inherit',
  },
  signIcon: {
    color: 'white',
  },
  stampLookupUrl: {
    position: 'absolute',
    whiteSpace: 'nowrap',
  },
  imageContainer: {
    display: 'flex',
    height: '100%',
    width: '100%',
    justifyContent: 'center',
    alignItems: 'center',
  },
  clearButton: {
    height: 16,
    width: 39,
    borderRadius: 3,
    border: `1px solid ${theme.palette.neutrals.tertiaryGrey}`,
    backgroundColor: 'white',
    fontSize: 9,
    animation: 'loaded 0.8s linear, unloaded 0.8s linear',
    zIndex: theme.zIndex.activeItem,
    position: 'absolute',
  },
  '@keyframes loaded': {
    '0%': { opacity: 0 },
    '25%': { opacity: '40%' },
    '50%': { opacity: '70%' },
    '75%': { opacity: '90%' },
    '100%': { opacity: '100%' },
  },
  '@keyframes unloaded': {
    '100%': { opacity: '100%' },
    '75%': { opacity: '90%' },
    '50%': { opacity: '70%' },
    '25%': { opacity: '40%' },
    '0%': { opacity: 0 },
  },
  image: {
    maxWidth: '100%',
    maxHeight: '100%',
    width: 'inherit',
    height: 'inherit',
  },
  meOnly: {
    backgroundColor: theme.palette.primary.main,
  },
});

export class InitialsC extends React.PureComponent<Props, State> {

  state = {
    hovered: false,
  };

  handleMouseEnter = () => {
    if (this.clearButtonTimeout) {
      clearTimeout(this.clearButtonTimeout);
    }
    this.setState({ hovered: true });
  };

  handleMouseLeave = () => {
    this.clearButtonTimeout = setTimeout(() => this.setState({ hovered: false }), 250);
  };

  clearButtonTimeout: TimeoutID;

  render() {
    const {
      classes, value, onClick, stampLookupUrl, wipe, loading, height, width, zoom, mode, disabled,
    } = this.props;

    const { hovered } = this.state;

    if (loading) {
      return (
        <div className={classNames(classes.root, classes.loaderContainer)}>
          <CircularProgress
            variant="indeterminate"
            value={100}
            className={classes.loader}
            size={24}
            thickness={2}
          />
        </div>
      );
    }

    if (!value) {
      return (
        <React.Fragment>
          <button
            type="button"
            className={classNames(
              classes.root,
              classes.signButton,
              // if field has no value, and it's a signing mode, add a background
              { [classes.meOnly]: mode === EDITOR_MODE.SIGNING },
            )}
            onClick={onClick}
          >
            <div className={classes.signIcon}>
              <InitialIcon />
            </div>
            <span className={classes.signText}>Initial</span>
          </button>
          <div
            className={classes.stampLookupUrl}
            style={{ top: height * zoom, fontSize: DEFAULT_STAMP_URL_FONT_SIZE * zoom }}
          >
            {STAMP_URL_PLACEHOLDER}
          </div>
        </React.Fragment>
      );
    }

    // Get svg image dimensions from svg string.
    const svgDimensions = getWidthHeightFromSVGString(value);
    let imageWidth = 0;
    let imageHeight = 0;

    if (svgDimensions) {
      const { width: svgWidth, height: svgHeight } = svgDimensions;
      const widthRatio = width * zoom / svgWidth;
      const heightRatio = height * zoom / svgHeight;

      // As a ratio, use the smallest one of two
      if (widthRatio > heightRatio) {
        imageWidth = svgWidth * heightRatio;
        imageHeight = svgHeight * heightRatio;
      } else {
        imageWidth = svgWidth * widthRatio;
        imageHeight = svgHeight * widthRatio;
      }
    }

    return (
      <React.Fragment>
        <div
          onMouseEnter={this.handleMouseEnter}
          onMouseLeave={this.handleMouseLeave}
          className={classes.imageContainer}
        >
          <img
            src={value}
            alt="initials"
            style={{
              maxHeight: '100%',
              maxWidth: '100%',
              width: imageWidth,
              height: imageHeight,
            }}
          />
        </div>
        <div
          className={classes.stampLookupUrl}
          style={{ top: height * zoom, fontSize: DEFAULT_STAMP_URL_FONT_SIZE * zoom }}
        >
          {stampLookupUrl}
        </div>
        {
          hovered && mode === EDITOR_MODE.SIGNING && !disabled
            && (
            <button
              type="button"
              className={classes.clearButton}
              onClick={wipe}
              onMouseEnter={this.handleMouseEnter}
              onMouseLeave={this.handleMouseLeave}
              style={{ left: `${width * zoom}px` }}
            >
              Clear
            </button>
            )
        }
      </React.Fragment>
    );
  }

}

export default withStyles(styles)(InitialsC);
