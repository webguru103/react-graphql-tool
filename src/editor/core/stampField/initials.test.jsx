import React from 'react';
import { render } from 'react-testing-library';
import { InitialsC } from './initials';

describe('Initial Field', () => {
  test('renders', () => {
    const { container } = render(<InitialsC classes={{}} />);
    expect(container).toMatchSnapshot();
  });
});
