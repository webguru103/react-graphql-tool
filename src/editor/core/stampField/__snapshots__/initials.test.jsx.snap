// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Initial Field renders 1`] = `
<div>
  <button
    class=""
    type="button"
  >
    <div>
      <svg
        height="20"
        viewBox="0 0 20 20"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
        xmlns:xlink="http://www.w3.org/1999/xlink"
      >
        <defs>
          <path
            d="M14.302 1.139h-.23a.569.569 0 0 1 0-1.139h3.362a2.077 2.077 0 0 1 1.458 3.558c.61.44 1.027 1.117 1.027 1.898 0 1.29-1.173 2.355-2.485 2.355H14.87a.569.569 0 0 1-.568-.569V1.14zm-2.547 5.564l-.477-1.09a.571.571 0 0 1-.103.009h-3.04l-.028-.001-.467 1.07a.57.57 0 0 1-.141 1.12H6.227a.569.569 0 0 1 0-1.139h.18L9.173.342a.568.568 0 0 1 1.042 0l2.767 6.33h.227a.569.569 0 0 1 0 1.139h-1.271a.569.569 0 0 1-.184-1.108zm-3.15-2.22h2.18L9.695 1.99l-1.09 2.493zm-5.866.36L3.93 3.65a.568.568 0 1 1 .805.805l-1.19 1.192 1.19 1.192a.57.57 0 0 1-.804.805L2.74 6.452 1.548 7.644a.568.568 0 1 1-.805-.805l1.191-1.192-1.19-1.192a.57.57 0 0 1 .804-.805l1.19 1.192zM.57 11a.569.569 0 0 1 0-1.138H20.43a.569.569 0 0 1 0 1.138H.57zm16.865-4.327c.708 0 1.348-.581 1.348-1.217 0-.694-.75-1.3-1.546-1.3H15.44v2.517h1.995zm0-3.655a.939.939 0 0 0 0-1.88h-1.995v1.88h1.995z"
            id="a"
          />
        </defs>
        <g
          fill="none"
          fill-rule="evenodd"
          transform="translate(0 5)"
        >
          <mask
            fill="#fff"
            id="b"
          >
            <use
              xlink:href="#a"
            />
          </mask>
          <use
            fill="currentColor"
            fill-rule="nonzero"
            xlink:href="#a"
          />
          <g
            mask="url(#b)"
          >
            <path
              d="M0-2h22v15H0z"
            />
          </g>
        </g>
      </svg>
    </div>
    <span>
      Initial
    </span>
  </button>
  <div
    style="font-size: NaNpx;"
  >
    dealtap.ca/xxx-xxx
  </div>
</div>
`;
